src/VisualStudio/Core/Def/Implementation/DebuggerIntelliSense/DebuggerIntellisenseFilter.cs

```diff
@@ -16,26 +17,63 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.DebuggerIntelliSense
 {
-    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable
+    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable, IFeatureController
         where TPackage : AbstractPackage<TPackage, TLanguageService>
         where TLanguageService : AbstractLanguageService<TPackage, TLanguageService>
     {
         private readonly ICommandHandlerServiceFactory _commandFactory;
         private readonly IWpfTextView _wpfTextView;
+        private readonly IFeatureServiceFactory _featureServiceFactory;
         private AbstractDebuggerIntelliSenseContext _context;
         private IOleCommandTarget _originalNextCommandFilter;
+        private IFeatureDisableToken _completionDisabledToken;
 
         internal bool Enabled { get; set; }
 
         public DebuggerIntelliSenseFilter(
             AbstractLanguageService<TPackage, TLanguageService> languageService,
             IWpfTextView wpfTextView,
             IVsEditorAdaptersFactoryService adapterFactory,
-            ICommandHandlerServiceFactory commandFactory)
+            ICommandHandlerServiceFactory commandFactory,
+            IFeatureServiceFactory featureServiceFactory)
             : base(languageService, wpfTextView, adapterFactory, commandFactory)
         {
             _wpfTextView = wpfTextView;
             _commandFactory = commandFactory;
+            _featureServiceFactory = featureServiceFactory;
+        }
+
+        internal void EnableCompletion()
+        {
+            var featureService = _featureServiceFactory.GetOrCreate(WpfTextView);
+            if (!featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
+            {
+                var completionDisabledToken = _completionDisabledToken;
+                if (completionDisabledToken == null)
+                {
+                    return;
+                }
+
+                completionDisabledToken.Dispose();
+                _completionDisabledToken = null;
+
+                var document = _context.ContextBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+                if (document == null)
+                {
+                    return;
+                }
+
+                _context.OpenDocumentAndLinkedDocuments(document);
+            }
+        }
+
+        internal void DisableCompletion()
+        {
+            var featureService = _featureServiceFactory.GetOrCreate(WpfTextView);
+            if (featureService.IsEnabled(PredefinedEditorFeatureNames.Completion) && _completionDisabledToken == null)
```

Comment:
If we get a call to Disable but somebody else had already temporarily disabled it for some other reason, shouldn't we still make a token?

Reply:
We cannot match pairs of disable and enable calls. So, if there are two disable calls in a row and then a single enable one, we should enable after. Therefore, do not need to disable if already disabled and do not need to enable if already enabled.